core/secrets/secret.go

```diff
@@ -86,3 +89,110 @@ func (c *SecretConfig) Validate() error {
 	}
 	return nil
 }
+
+// URL represents a reference to a secret.
+type URL struct {
+	Version        string
+	ControllerUUID string
+	ModelUUID      string
+	Path           string
+	Attribute      string
+}
+
+/*
+Example secret URLs:
+	secret://v1/apitoken
+	secret://v1/mariadb.dbpass
+	secret://v1/apache.catalog.password
+	secret://v1/proxy#key
+	secret://v1/cfed9630-053e-447a-9751-2dc4ed429d51/myawscredential
+	secret://v1/cfed9630-053e-447a-9751-2dc4ed429d51/11111111-053e-447a-6666-2dc4ed429d51/myawscredential
+*/
+
+var uuidSnippet = `[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}`
+
+var secretURLParse = regexp.MustCompile(`^` +
+	fmt.Sprintf(
+		`((?P<version>v[1-9])\/)((?P<controllerUUID>%s)\/)?((?P<modelUUID>%s)\/)?(?P<path>[0-9a-zA-Z]+(\.[0-9a-zA-Z]+)*)(#[0-9a-zA-Z]+)?`,
+		uuidSnippet, uuidSnippet) +
+	`$`)
+
+// ParseURL parses the specified URL string into a URL.
+func ParseURL(str string) (*URL, error) {
+	u, err := url.Parse(str)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if u.Scheme != "secret" {
+		return nil, errors.NotValidf("secret URL scheme %q", u.Scheme)
+	}
+	spec := fmt.Sprintf("%s%s", u.Host, u.Path)
+	if !secretURLParse.MatchString(spec) {
+		return nil, errors.NotValidf("secret URL %q", str)
+	}
+	result := &URL{
+		Version:        secretURLParse.ReplaceAllString(spec, "$version"),
+		ControllerUUID: secretURLParse.ReplaceAllString(spec, "$controllerUUID"),
+		ModelUUID:      secretURLParse.ReplaceAllString(spec, "$modelUUID"),
+		Path:           secretURLParse.ReplaceAllString(spec, "$path"),
+		Attribute:      u.Fragment,
+	}
+	return result, nil
+}
+
+// ShortString prints the URL without controller or model UUID.
+func (u *URL) ShortString() string {
+	fullPath := []string{"secret:/", u.Version}
+	fullPath = append(fullPath, u.Path)
+	str := strings.Join(fullPath, "/")
+	if u.Attribute != "" {
+		str += "#" + u.Attribute
+	}
+	return str
+}
+
+// String prints the URL as a string.
+func (u *URL) String() string {
+	fullPath := []string{"secret:/", u.Version}
+	if u.ControllerUUID != "" {
+		fullPath = append(fullPath, u.ControllerUUID)
+	}
+	if u.ModelUUID != "" {
+		fullPath = append(fullPath, u.ModelUUID)
+	}
+	fullPath = append(fullPath, u.Path)
+	str := strings.Join(fullPath, "/")
+	if u.Attribute != "" {
+		str += "#" + u.Attribute
+	}
+	return str
+}
+
+// SecretMetadata holds metadata about a secret.
+type SecretMetadata struct {
+	// Read only after creation.
+	Path  string
+	Scope Scope
+
+	// Version starts at 1 and is incremented
+	// whenever an incompatible change is made.
+	Version int
```

Comment:
Maybe uint instead of int to avoid negative numbers

Reply:
I used int because it avoids casting in various places. But I changed to uint.